File: InterfacageNumerique/TP/bloc_rayonnement/codes/test_comm_led.cpp
/*  
 *  Structure of a main file for embedded project @ LEnsE
 *      Light Radiation Project / Communication
 *
 *  This code allows communication between a computer application
 * and a Nucleo board. 
 *  Sending an 'a' on the serial port @115200 bauds ... the LED2.
 *  Sending an 'e' on the serial port @115200 bauds ... the LED2.
 *****************************************************************
 *  Pinout :
 *      PC_7 - Output LED (D1 on the light radiation board)
 *      PB_13 - Output LED (D2 on the light radiation board)
 *****************************************************************
 *  Tested with Nucleo XnnnMM board / Mbed OS 6
 *****************************************************************
 *  Author : J. VILLEMEJANE / LEnsE - Creation 2025/01/18
 *****************************************************************
 *  LEnsE / https://lense.institutoptique.fr/
 *      Based on Mbed OS 6 example : mbed-os-example-blinky-baremetal
 */

#include "mbed.h"

DigitalOut led1(PC_7);
DigitalOut led2(PB_13);

UnbufferedSerial      my_pc(USBTX, USBRX);

void ISR_my_pc_reception(void);

char data;

int main()
{    
    my_pc.baud(115200);
    my_pc.attach(&ISR_my_pc_reception, UnbufferedSerial::RxIrq);
    while (true){}
}
void ISR_my_pc_reception(void){
    led1 = !led1;
    my_pc.read(&data, 1);     // get the received byte
    if(data == 'a'){    led2 = 1;   } 
    if(data == 'e'){    led2 = 0;   }    
    my_pc.write(&data, 1);    // echo of the byte received
}

File: InterfacageNumerique/TP/bloc_rayonnement/codes/test_comm_led.py
# -*- coding: utf-8 -*-
"""
First code to transmit and receive data between a computer and
a Nucleo Board.

***************************************************************
  Tested with Nucleo L476RG board / Arduino + STM32
***************************************************************

Creation 2024/12/16
@author: julien.villemejane <julien.villemejane@institutoptique.fr>
LEnsE / https://lense.institutoptique.fr/
"""

from serial import Serial
import serial.tools.list_ports


def receiving_data(ser):
    received_data = ""
    while True:
        byte = ser.read(1).decode('ascii')
        if byte == ';':  # Check if byte is ending character
            break
        received_data += byte
    return received_data


if __name__ == "__main__":
    ports = serial.tools.list_ports.comports()
    # To obtain the list of the communication ports
    for port, desc, hwid in sorted(ports):
        print("{}: {}".format(port, desc))
    # To select the port to use
    selectPort = input("Select a COM port : ")    
    print(f"Port Selected : COM{selectPort}")
    # To open the serial communication at a specific baudrate
    serNuc = Serial('COM'+str(selectPort), 9600)

    appOk = 1

    while appOk:
        choice = input("Char to send : ") 
        if choice == 'q' or choice == 'Q':
            appOk = 0
        else:
            if choice == 'a': # Send 'a'
                data_to_send = 'a'
                serNuc.write(bytes(data_to_send, 'ascii'))
                data_rec = receiving_data(serNuc)
                print(str(data_rec))
            elif choice == 'e': # Send 'e'
                data_to_send = 'e'
                serNuc.write(bytes(data_to_send, 'ascii'))
                data_rec = receiving_data(serNuc)
                print(str(data_rec))            
       
    
    
    serNuc.close()
    